packages/devtools_app/lib/src/screens/memory/panes/diff/controller/heap_diff.dart

```diff
@@ -56,35 +58,105 @@ class _HeapCouple {
+  late final Map<HeapClassName, DiffClassStats> classesByName;
+  late final List<DiffClassStats> classes =
+      classesByName.values.toList(growable: false);
 
-  final _HeapCouple heapCouple;
+  @override
+  void seal() {
+    super.seal();
+    for (var c in classes) {
+      c.seal();
+    }
+  }
+}
 
-  HeapStatistics _stats() {
-    final result = <String, HeapClassStatistics>{};
+class DiffClassStats extends ClassStats {
+  DiffClassStats._({
+    required this.heapClass,
+    required this.total,
+    required StatsByPath objectsByPath,
+  }) : super(objectsByPath);
+
+  final HeapClassName heapClass;
+  final ObjectSetDiff total;
+
+  static DiffClassStats? diff({
+    required SingleClassStats? before,
+    required SingleClassStats? after,
+  }) {
+    if (before == null && after == null) return null;
+
+    final heapClass = (before?.heapClass ?? after?.heapClass)!;
+
+    final result = DiffClassStats._(
+      heapClass: heapClass,
+      total: ObjectSetDiff(before: before?.objects, after: after?.objects),
+      objectsByPath: subtractMaps<ClassOnlyHeapPath, ObjectSetStats,
+          ObjectSetStats, ObjectSetStats>(
+        from: after?.statsByPath,
+        substract: before?.statsByPath,
+        subtractor: ({subtract, from}) =>
+            ObjectSetStats.subtract(subtract: subtract, from: from),
+      ),
+    );
+
+    if (result.isZero()) return null;
+    return result..seal();
+  }
 
-    final older = heapCouple.older.stats.statsByClassName;
-    final younger = heapCouple.younger.stats.statsByClassName;
+  bool isZero() => total.isZero;
+}
 
-    final unionOfKeys = older.keys.toSet().union(younger.keys.toSet());
+class ObjectSetDiff {
+  ObjectSetDiff({ObjectSet? before, ObjectSet? after}) {
+    before ??= ObjectSet.empty;
+    after ??= ObjectSet.empty;
 
-    for (var key in unionOfKeys) {
-      final olderStats = older[key];
-      final youngerStats = younger[key];
+    final codesBefore = before.objectsByCodes.keys.toSet();
+    final codesAfter = after.objectsByCodes.keys.toSet();
 
-      if (olderStats != null && youngerStats != null) {
-        final diff = HeapClassStatistics.subtract(olderStats, youngerStats);
-        if (!diff.isZero) result[key] = diff;
-      } else if (youngerStats != null) {
-        result[key] = youngerStats;
-      } else {
-        result[key] = HeapClassStatistics.negative(olderStats!);
+    final allCodes = codesBefore.union(codesAfter);
+    for (var code in allCodes) {
+      if (codesBefore.contains(code) && (codesAfter.contains(code))) continue;
+
+      final object = before.objectsByCodes[code] ?? after.objectsByCodes[code]!;
+
+      if (codesBefore.contains(code)) {
+        deleted.countInstance(object);
+        delta.uncountInstance(object);
+        continue;
+      }
+      if (codesAfter.contains(code)) {
+        created.countInstance(object);
+        delta.countInstance(object);
+        continue;
       }
+      assert(false);
```

Comment:
do we need this? this is unreachable

Reply:
If somebody make a change in the code above, and this place accidentally will become reachable, I want tests to fail.
And, this assert improves readability -  confirmation that all cases are covered.

(BTW, the code is not marked as unreachable by IDE)